### PR TITLE
Add support for `OrderedDict` to `unpack_remotedata`

### LINKED ISSUE #5281: Dask futures nested in `OrderedDict` types are not resolved when using `submit`
**What happened**:

I'm passing a dask future to another delayed function call using `client.submit`. If the dask future is in a dictionary, it is resolved to a value when it reaches the function. If it is in an ordered dictionary, it remains a dask future.

**What you expected to happen**:

Futures nested in `OrderedDict` collections should be resolved to values

**Minimal Complete Verifiable Example**:

```python
from dask.distributed import Client
from collections import OrderedDict


def identity(x):
    return x


def show(x):
    print(repr(x))


if __name__ == "__main__":
    with Client() as client:
        f1 = client.submit(show, "hello")
        f2 = client.submit(show, client.submit(identity, "hello"))

        # Works
        f3 = client.submit(show, {"key": client.submit(identity, "hello")})
        
        # Does not work
        f4 = client.submit(show, OrderedDict({"key": client.submit(identity, "hello")}))

        _ = [f.result() for f in (f1, f2, f3, f4)]
```

```
'hello'
'hello'
{'key': 'hello'}
OrderedDict([('key', <Future: pending, key: identity-36cc7da4255c87ef48c35a6af40f990f>)])
```

**Anything else we need to know?**:

`unpack_collections` appears to support this, but I don't think it's used here.

```python
...
elif typ in (dict, OrderedDict):
    tsk = (typ, [[_unpack(k), _unpack(v)] for k, v in expr.items()])
```

**Environment**:

- Dask version:  2021.8.1+18.gda2bcccc / master
- Python version: 3.7
- Operating System: OSX
- Install method (conda, pip, source): Source (also tested via pip)

### PR DESCRIPTION
**Pull request by madkinsz**
- [x] Closes #5281
- [x] Tests added / passed
- [x] Passes `black distributed` / `flake8 distributed` / `isort distributed`

I don't see any tests for `unpack_remotedata` nor tests covering future unpacking in `test_highlevelgraph` -- Anywhere you'd like me to add a test case for this?

### CODE REVIEW

File: distributed/utils_comm.py [hunk 1]
```diff
@@ -233,7 +233,7 @@ def unpack_remotedata(o, byte_keys=False, myset=None):
     elif typ in (dict, OrderedDict):
         if not o:
             return o
-        return typ([[k, unpack_remotedata(v, byte_keys, myset)] for k, v in o.items()])
+        return typ((k, unpack_remotedata(v, byte_keys, myset)) for k, v in o.items())
```
[the remark this answers is not in the document]
Comment (jrbourbeau) [reply distributed/utils_comm.py:236]: That's a fair question. Let's keep things as they are and modify later if needed

File: distributed/tests/test_utils_comm.py [hunk 2]
```diff
@@ -143,3 +151,13 @@ async def my_sleep(amount):
...
+
+
+@pytest.mark.parametrize("dict_type", [dict, OrderedDict])
+def test_unpack_remotedata_unpacks_dicts(dict_type):
+    rd1 = WrappedKey("key-1")
+    rd2 = WrappedKey("key-2")
+    assert unpack_remotedata(dict_type([(1, rd1), (2, rd2)])) == (
+        dict_type([(1, "key-1"), (2, "key-2")]),
+        {rd1, rd2},
+    )
```
Comment (jrbourbeau) [inline distributed/tests/test_utils_comm.py:163]: Thanks you for adding this lower-level test for `unpack_remotedata`. Could you also add a higher-level test to `distributed/tests/test_client.py` which ensure the user-facing result from the original issue is as expected? For example, we could add an `assert` that

```python
client.submit(show, OrderedDict({"key": client.submit(identity, "hello")}))
```

(or something similar) is returning the correct result.

Comment (madkinsz) [reply distributed/tests/test_utils_comm.py:163]: 👍 will do

Comment (madkinsz) [reply distributed/tests/test_utils_comm.py:163]: This file doesn't seem to have any tests of that sort (but there are a lot of tests in there). Should I just put it at the bottom or is there somewhere else in there that makes sense to you ?

Comment (jrbourbeau) [reply distributed/tests/test_utils_comm.py:163]: No strong preference, wherever you prefer in that module

Comment (madkinsz) [reply distributed/tests/test_utils_comm.py:163]: I added a test but it's failing for `OrderedDict` -- looks like there's somewhere else that needs changes :( good call on adding a client level test!

**Comment by madkinsz** [comment]
This is not working and is stale, so I'm going to close it. Feel free to ping me if you're continuing this work and need anything.

**Pull request closed without merging**